service/raft/group0_state_machine.cc

```diff
@@ -165,6 +165,13 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
         co_await write_mutations_to_database(_sp, cmd.creator_addr, std::move(chng.mutations));
         co_await _ss.topology_transition();
     },
+    [&] (mixed_change& chng) -> future<> {
+        auto topo_mut = std::vector<canonical_mutation>(chng.mutations.begin(), chng.mutations.begin()+chng.topo_muts_count);
+        auto schema_mut = std::vector<canonical_mutation>(chng.mutations.begin()+chng.topo_muts_count, chng.mutations.end());
```

Comment:
Use move iterators to avoid copies
https://en.cppreference.com/w/cpp/iterator/make_move_iterator

Reply:
But wouldn't it be better to change the struct definition to have two vectors?

Reply:
> But wouldn't it be better to change the struct definition to have two vectors?

plz see https://github.com/scylladb/scylladb/pull/16723#discussion_r1580082294



> Use move iterators to avoid copies https://en.cppreference.com/w/cpp/iterator/make_move_iterator

Is it safe to "drain" the vector here by using move iterators? I know that the current code looks silly, I am leaving it like that until we agree if I should continue with a single or 2 vectors (to not change it back and forth).

Reply:
Well, in our latest discussion (me, you, and @tgrabiec) on one of the tablets meeting, the conclusion was that the entire vector of mutations (which includes both schema and topology mutations) should be merged inside a single `merge_schema_from` call because we want to merge these mutations atomically.

Which means you don't need a separate `write_mutations_to_database` call. Which means you don't need two vectors. So IIUC you don't need either `topo_muts_count` or storing two vectors inside `mixed_change`, you just need a single vector with everything without having to know which mutations are for schema and which for topology.

> Is it safe to "drain" the vector here by using move iterators?

Are you accessing the vector after this code?

Reply:
Are you saying I can revert the patch that introduced a new raft cmd and just use `schema_change`?
Or do I still need to wait for:
```
        co_await _ss.topology_transition();
```
I introduced `mixed_change` because topo sm state transition was not reflected in memory and I was forever stuck in a never ending `rf_change` state.

Reply:
You still need `co_await _ss.topology_transition();` to refresh the in-memory state. So you still need a new command. You just don't need two vectors or a count to separate the different mutations.